### PR TITLE
Window Server 2022 Auto Install Authoring

### PR DESCRIPTION
**Pull request by bobsira**
This pull request introduces automation scripts for setting up a virtual machine (VM) with Windows Server 2022 and configuring it for Minikube. The changes include a PowerShell script for VM creation and an unattended XML file for automated OS installation and configuration.

### VM Setup Automation:
* Added `SetUp.ps1` script to automate the creation of a Hyper-V VM with predefined settings, including memory allocation, virtual hard disk (VHD) configuration, and boot device setup. The script also configures the VM to use virtualization extensions and mounts necessary ISO files.

### Unattended OS Installation and Configuration:
* Added `autounattend.xml` file to automate Windows Server 2022 installation. It includes configurations for disk partitioning, locale settings, and OS image installation.
* Configured the system to enable WinRM over HTTPS with a self-signed certificate, add firewall rules, and enable Basic authentication for remote management.
* Set up first logon commands to install the Containers feature, disable the Server Configuration tool (SConfig), and restart the system to apply changes.
* Configured user accounts with an administrator password and enabled auto-login for the Administrator account.To test the PR:

**Edit your Windows Server 2022 ISO**: You can use free software like AnyBurn or ImgBurn

**Add the `autounattend.xml` file**: Copy your `autounattend.xml` file to the root folder path of your  Windows Server 2022 ISO file.

**Create a new ISO**: Create a new ISO with the autounattend.xml file bundled together.

**Run SetUp.ps1 file**: Remember to replace the $SwitchName and $ISOFile with the appropriate values

### CODE REVIEW

File: automation/SetUp.ps1 [hunk 1]
```diff
@@ -0,0 +1,17 @@
+$SwitchName = "External VM Switch"
```
Comment (iankingori) [inline automation/SetUp.ps1:1]: can all these 3 be parameters passed to the script with a default value?

File: automation/SetUp.ps1 [hunk 2]
```diff
@@ -0,0 +1,17 @@
+$SwitchName = "External VM Switch" 
+$ISOFile = "C:\Users\bosira\Downloads\SERVER_EVAL_x64FRE_en-us-uni.iso"
```
Comment (iankingori) [inline automation/SetUp.ps1]: remove the bosira bit amd use `~` or `$HOME`

File: automation/autounattend.xml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                </SynchronousCommand>
+                <SynchronousCommand wcm:action="add">
+                    <CommandLine>%SystemRoot%\system32\WindowsPowerShell\v1.0\powershell.exe -Command Restart-Computer -Force</CommandLine>
+                    <Order>7</Order>
+                    <Description>Restart computer to apply changes</Description>
+                </SynchronousCommand>
+            </FirstLogonCommands>
+            <UserAccounts>
+                <AdministratorPassword>
+                    <Value>M@kindu.2023</Value>
```
Comment (iankingori) [inline automation/autounattend.xml]: use a more generic or context appropriate default password e.g Minikube@2022?

**Comment by mloskot** [comment]
Is there any way to avoid re-burning the ISO?

By the way, I've been [experimenting with official Windows Server VHD](https://github.com/kubernetes-sigs/sig-windows-dev-tools/blob/3783f14b24b0fac9fae4715e1f527820412c467e/experiments/swdt/samples/mloskot/README.windows.md) instead of ISO. It still does not offer full automation, but needs less interactive steps (password, language, eula).

/cc @knabben This is interesting. If there was portable CLI ISO burner that could be scripted :)

**Comment by vrapolinario** [comment]
I'd also recommend moving away from burning an ISO. The goal here is to make the process easier for the user, and this adds more complexity.

File: automation/SetUp.ps1 [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    NewVHDPath = "${env:homepath}\.minikube\machines\$VMName\VHD.vhdx";
+    NewVHDSizeBytes = 10GB;
+    BootDevice = "VHD";
+    Path = "${env:homepath}\.minikube\machines\";
+    SwitchName = $SwitchName
+}
+New-VM @VM
+Set-VM -Name $VMName -ProcessorCount 2 -AutomaticCheckpointsEnabled $false
+Set-VMProcessor -VMName $VMName -ExposeVirtualizationExtensions $true
+Set-VMDvdDrive -VMName $VMName -Path $ISOFile
```
Comment (iankingori) [inline automation/SetUp.ps1:16]: I can't push changes to your fork but here's a suggestion that has worked for me to work around adding the answer file to the windows iso.

I've created a disk image that contains just the answer file and added a new dvd drive to the machine. Windows by default searches other disk drives for an unattended answer files meaning we can use default installation media and only update this iso if the answer file changes

`Add-VMDvdDrive -VMName $VMName -Path "$PSScriptRoot\auto-install.iso" -ControllerNumber 1 -ControllerLocation 1`

Comment (iankingori) [reply automation/SetUp.ps1:16]: See this PR: https://github.com/bobsira/MinikubeWindowsContainers/pull/1/files

File: automation/autounattend.xml [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+            <!-- https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-international-core -->
+            <InputLocale>en-US</InputLocale>
+            <SystemLocale>en-US</SystemLocale>
+            <UILanguage>en-US</UILanguage>
+            <UILanguageFallback>en-US</UILanguageFallback>
+            <UserLocale>en-US</UserLocale>
+        </component>
+        <component name="Microsoft-Windows-Shell-Setup" processorArchitecture="amd64" publicKeyToken="31bf3856ad364e35" language="neutral" versionScope="nonSxS" xmlns:wcm="http://schemas.microsoft.com/WMIConfig/2002/State" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+            <!-- https://docs.microsoft.com/en-us/windows-hardware/customize/desktop/unattend/microsoft-windows-shell-setup -->
+            <ComputerName>minikube-m05</ComputerName>
```
Comment (iankingori) [inline automation/autounattend.xml]: naming the computer here means the vm-name passed to the create script and the computer name do not match if I use something like minikube-03.

It should be possible to drop this and use powershell to set the computername after startup

File: automation/SetUp.ps1 (hunk 4 above)
Comment (bobsira) [reply automation/SetUp.ps1:16]: I've approved this PR and merged your changes.

File: automation/autounattend.xml (hunk 5 above)
Comment (bobsira) [reply automation/autounattend.xml]: It should be possible to rename the computer even with this entry here. We should maybe come up with a better place holder name in the event someone does not want to bother renaming the computer name

File: automation/SetUp.ps1 (hunk 1 above)
Comment (iankingori) [reply automation/SetUp.ps1:1]: I'd like to execute this using something like: `Setup.ps1 -SwitchName "External VM Switch" -ISOFilePath "~\Downloads\..." -VMName "minikube-windows-1"`

**Comment by bobsira** [comment]
The changes in this PR has been merged in the remoting PR found here: https://github.com/vrapolinario/MinikubeWindowsContainers/pull/4

**Comment by bobsira** [comment]
We are closing this PR in favor of this https://github.com/vrapolinario/MinikubeWindowsContainers/pull/4

**Comment by bobsira** [comment]
> I'd also recommend moving away from burning an ISO. The goal here is to make the process easier for the user, and this adds more complexity.

This has been addressed with help from @iankingori

**Pull request closed without merging**